### PR TITLE
SGEMM: fix MT version issues

### PR DESCRIPTION
**Pull request by jzielins**
This commit adds additional validation of input arguments (matrix sizes) to prevent errors in algorithms that require matrix size to be a multiple of SIMD width. It also modifies the mechanism of

### CODE REVIEW

File: examples/sgemm/SGEMM_kernels.ispc [hunk 1]
```diff
@@ -85,8 +94,9 @@ static task void SGEMM_naive_task(uniform float matrixA[], uniform float matrixB
 
 export void SGEMM_naive_withTasks(uniform float matA[], uniform float matB[], uniform float matC[],
                                     uniform int M, uniform int N, uniform int K) {
-    // Generate more tasks than threads, to enable some load balancing:
-    uniform int numTasks = num_cores() * 4;
+    // Task must have at least programCount columns to process, but we require the matrix
+    // K size to be a miltiple of programCount, so it's safe to divide here
+    uniform int numTasks = M / programCount;
```
Comment (dbabokin) [inline examples/sgemm/SGEMM_kernels.ispc:99]: Either comment or code is wrong. Comment talks about K, code uses M.

File: examples/sgemm/SGEMM_kernels.ispc [hunk 2]
```diff
@@ -403,7 +412,7 @@ static task void SGEMM_tileNoSIMDIntrin_task(uniform float matrixA[], uniform fl
 export void SGEMM_tileNoSIMDIntrin_withTasks(uniform float matA[], uniform float matB[], uniform float matC[],
     uniform int M, uniform int N, uniform int K) {
     // Generate more tasks than threads, to enable some load balancing:
-    uniform int numTasks = num_cores()*4;
+    uniform int numTasks = M / programCount;
```
Comment (dbabokin) [inline examples/sgemm/SGEMM_kernels.ispc:415]: Some use ``K / programCount``, some use ``M / programCount``. I think it requires a note in comments to explain this.

File: examples/sgemm/SGEMM_main.cpp [hunk 3]
```diff
@@ -231,7 +232,18 @@ int main(int argc, char **argv) {
...
         printf("%s\n", argv[0]);
         printf("\nInvalid number of inputs\n");
-        getchar();
+        exit(-1);
+    }
+
+    int programCount = SGEMM_get_program_count();
+    int tileSize = SGEMM_get_tile_size();
+    if (K % programCount != 0 || K % tileSize != 0) {
+        printf("\nNumber of columns in Matrix B must be a multiple of %d and %d!\n", programCount, tileSize);
```
Comment (dbabokin) [inline examples/sgemm/SGEMM_main.cpp:241]: Let's have additional clarification what these numbers are. i.e.
``"...of %d (target width) and %d (tile size)!\n"``

File: examples/sgemm/SGEMM_main.cpp [hunk 4]
```diff
@@ -231,7 +232,18 @@ int main(int argc, char **argv) {
...
+        exit(-1);
+    }
+
+    if (M % programCount != 0) {
+        printf("\nNumber of rows in Matrix A must be a multiple of %d!\n", programCount);
```
Comment (dbabokin) [inline examples/sgemm/SGEMM_main.cpp:246]: Same here:
``"... of %d (target width)!\n"``

File: examples/sgemm/SGEMM_main.cpp [hunk 5]
```diff
@@ -243,7 +255,7 @@ int main(int argc, char **argv) {
     init_matrix_rand(matrixB, N, K, 10.0f);
     float *matrixC;
     matrixC = (float *)malloc(M * K * sizeof(float));
-    init_matrix_rand(matrixC, M, K, 0.0f);
```
Comment (dbabokin) [inline examples/sgemm/SGEMM_main.cpp:246]: why is this change?

File: examples/sgemm/SGEMM_kernels.ispc (hunk 1 above)
Comment (jzielins) [reply examples/sgemm/SGEMM_kernels.ispc:99]: Ctrl-C, Ctrl-V error, fixed.

Comment (jzielins) [reply examples/sgemm/SGEMM_kernels.ispc:99]: Ctrl-C, Ctrl-V error, fixed.

File: examples/sgemm/SGEMM_main.cpp (hunk 3 above)
Comment (jzielins) [reply examples/sgemm/SGEMM_main.cpp:241]: Done.

File: examples/sgemm/SGEMM_main.cpp (hunk 5 above)
Comment (jzielins) [reply examples/sgemm/SGEMM_main.cpp:246]: Code cleanup. I thought that using rand with limit 0.0f is a little bit of overkill.  init_matrix(..., value) seems more logical here, although functionally it's the same thing. Additionally, the init_matrix() function was not used in the code, so we should either remove or use it.

File: examples/sgemm/SGEMM_main.cpp (hunk 4 above)
Comment (jzielins) [reply examples/sgemm/SGEMM_main.cpp:246]: Good point, fixed

File: examples/sgemm/SGEMM_kernels.ispc (hunk 2 above)
Comment (jzielins) [reply examples/sgemm/SGEMM_kernels.ispc:415]: True. Fixed.

File: examples/sgemm/SGEMM_main.cpp (hunk 5 above)
Comment (dbabokin) [reply examples/sgemm/SGEMM_main.cpp:246]: Ah, you are right! We definitely need to use init_matrix() here.

**Pull request merged**